Add printer config modal

Cellivar opened the pull request:
Adds a modal dialog for configuring printer details. The printer config is limited to relevant config items that might need to be changed or validated at the con. 

This wires the previously added config buttons that weren't doing anything.

![image](https://user-images.githubusercontent.com/1441553/148000343-f5ec98f5-95c5-4d0d-b1b5-1796cb7e546d.png)

Cellivar committed:
Add printer config modal

Cellivar committed:
Merge branch 'main' into feature/add-printer-config-page

kobaj commented on public/regdesk.html:
```diff
@@ -199,7 +209,77 @@ <h4>Payment required</h4>
+              </div>
+              <div class="col-md-4">
+                <label for="printerSettingsDarkness" class="col-sm-6 col-form-label">Darkness</label>
+                <select id="printerSettingsDarkness" class="form-select" aria-label="Darkness">
```
You might put some kind of human connotation to these options.

```
<option value=1>darkest</option>
<option>2</option>
...
<option value=15>lightest</option>
```

kobaj commented on public/regdesk.html:
```diff
@@ -199,7 +209,77 @@ <h4>Payment required</h4>
+            <div class="row">
+              <div class="col-md-12 alert alert-danger">
+                Do not change the configuration when the label dispenser is engaged. It will dispense many labels and
+                not stop. Ensure dispenser is disengaged before saving.
```
What does this mean? There isn't a button called "disengage". Is a human supposed to interpret that to mean "while printing"? Perhaps it would be simpler to say "while printing".

Cellivar replied:
This is the boundary where we expect someone to know what they're doing when reading this. This page here will feature in the help documentation I've started writing that will be in a physical folder with the printers.

![image](https://user-images.githubusercontent.com/1441553/148002150-6b3cab6c-6a5a-433f-95ea-d6df66851f7c.png)

kobaj commented on public/regdesk.html:
```diff
@@ -199,7 +209,77 @@ <h4>Payment required</h4>
+              </div>
+              <div class="col-md-4">
+                <label for="printerSettingsSpeed" class="col-sm-5 col-form-label">Speed</label>
+                <select id="printerSettingsSpeed" class="form-select" aria-label="Speed">
```
Speed of what? Printing? Label Dispensing? Data transfer?

Cellivar replied:
Good point, "Print Speed" it is.

kobaj commented on public/regdesk.html:
```diff
@@ -199,7 +209,77 @@ <h4>Payment required</h4>
+              <div class="col-md-4">
+                <label for="printerSettingsSpeed" class="col-sm-5 col-form-label">Speed</label>
+                <select id="printerSettingsSpeed" class="form-select" aria-label="Speed">
+                  <option>1</option>
```
Again, human connotation would be helpful.

```
<option value=1>fastest</option>
<option>2</option>
...
<option value=4>slowest</option>
```

kobaj commented on src/regdesk/printer_config_modal.js:
```diff
@@ -0,0 +1,64 @@
+
+    this.printer.labelWidth = (parseFloat(this.printerLabelWidth.value));
+    this.printer.labelHeight = (parseFloat(this.printerLabelHeight.value));
+    await this.printer.configDensity(this.printerDarkness.value);
+    await this.printer.configSpeed(this.printerSpeed.value);
```
Perhaps we should parseInt?

Cellivar replied:
Not necessary here for the density or speed, as the values are enumerated in the selects.

kobaj commented on src/regdesk/printer_config_modal.js:
```diff
@@ -0,0 +1,64 @@
+    await this.printer.configDensity(this.printerDarkness.value);
+    await this.printer.configSpeed(this.printerSpeed.value);
+    await this.printer.configPrintDirection();
+    await this.printer.configLabelWidth();
```
This method doesn't take the width as an argument?

Cellivar replied:
Nope, it updates the printer with whatever you set the width to previously, in this case on line 56. I plan to revisit the public interface on my WebZLP library at some point to clean up some of the oddities I've run into while using it.

Cellivar committed:
Make printer settings more human readable

Cellivar commented:
I've been working with these printers so much I'm blind to their settings being unparseable for anyone else, good call on that. I've updated them to be the real inches per second value and added hints to the darkness.

kobaj commented on public/regdesk.html:
```diff
@@ -240,9 +240,9 @@ <h5 class="modal-title">Printer Settings For: <span id="printerSettingsSerial"><
                 <select id="printerSettingsDarkness" class="form-select" aria-label="Darkness">
-                  <option>1</option>
+                  <option value="1">1 - Lightest</option>
                   <option>2</option>
```
Only just realizing, should these have a default value too?

Cellivar replied:
The value should always be pulled from the printer configuration and that internally sets a default, I don't think we need one here.

kobaj commented on public/regdesk.html:
```diff
@@ -240,9 +240,9 @@ <h5 class="modal-title">Printer Settings For: <span id="printerSettingsSerial"><
               <div class="col-md-4">
-                <label for="printerSettingsDarkness" class="col-sm-6 col-form-label">Darkness</label>
+                <label for="printerSettingsDarkness" class="form-label">Print Darkness</label>
                 <select id="printerSettingsDarkness" class="form-select" aria-label="Darkness">
```
Are these settings retrievable in some form? So that someone can read how a printer is currently configured when they pull up this screen?

Cellivar replied:
This interface would be the way to find them, yes.

kobaj reviewed (approved):
Totally forgot to include two major comments this time. My bad. They could be added in a later PR though. :)
